R/lint.R

```diff
@@ -0,0 +1,143 @@
+# Retrieve the lifecycle status defined in each Rd file
+db_lifecycle <- function(db) {
+  lifecycle_patterns <- paste0("(?:",
+    paste(collapse = "|",
+      c("lifecycle::badge\\([\\\\]\"",
+        "rlang:::lifecycle\\([\\\\]\"",
+        "list\\(\"lifecycle-",
+        "https://www.tidyverse.org/lifecycle/#"
+      )),
+    ")([\\w-]+)"
+  )
+
+  desc <- lapply(db, asNamespace("tools")$.Rd_get_metadata, "description")
+  lapply(desc, function(x) {
+    res <- regexpr(lifecycle_patterns, x, perl = TRUE)
+    starts <- attr(res, "capture.start")
+    ends <- starts + attr(res, "capture.length") - 1
+    substring(x, starts, ends)
+  })
+}
+
+# Retrieve the functions listed in the usage for each Rd file in the database
+db_function <- function(db) {
+  usage <- lapply(db, asNamespace("tools")$.Rd_get_section, "usage")
+  lapply(usage, get_usage_function_names)
+}
+
+#' @param package The name of installed packages.
+#' @param which The lifecycle statuses to return for installed packages.
+#'   Include `NA` if you want to include functions without a specified lifecycle
+#'   status in the results.
+#' @export
+#' @rdname lint_lifecycle
+pkg_function_lifecycle <- function(package, which = c("superseded", "deprecated", "questioning", "defunct", "experimental", "soft-deprecated", "retired")) {
+  rlang::check_installed("vctrs")
+  which <- match.arg(which, several.ok = TRUE)
+  stopifnot(is.character(package) && length(package) == 1)
+
+  db <- tools::Rd_db(package)
+  lc <- db_lifecycle(db)
+  funs <- db_function(db)
+
+  res <- mapply(function(lc, f) data.frame(fun = f, lifecycle = rep(lc, length(f)), stringsAsFactors = FALSE), lc, funs, SIMPLIFY = FALSE)
+
+  res <- vctrs::vec_rbind(!!!res)
+
+  # Filter funs without a lifecycle
+  if (!NA %in% which) {
+    res <- res[!is.na(res$lifecycle), ]
+  }
+
+  # Filter funs without a function name
+  res <- res[nzchar(res$fun), ]
+
+  # Filter method definitions
+  res <- res[grep("[\\\\]method\\{", res$fun, invert = TRUE), ]
+
+  # filter lifecycles not in which
+  res <- res[res$lifecycle %in% which, ]
+
+  if (nrow(res) == 0) {
+    return(data.frame(package = character(), fun = character(), lifecycle = character()))
+  }
+
+  res$package <- package
+
+  res[c("package", "fun", "lifecycle")]
+}
+
+get_usage_function_names <- function(x) {
+  if (!length(x)) {
+    character(1)
+  } else {
+    res <- asNamespace("tools")$.parse_usage_as_much_as_possible(x)
+    vapply(res, function(x) {
+      if (is.call(x)) as.character(x[[1]]) else character(1)
+    }, character(1))
+  }
+}
+
+#' Lint usages of functions that have a non-stable life cycle.
+#'
+#' - `lint_lifecycle` dynamically queries the package documentation for packages
+#'   in `packages` for lifecycle annotations and then searches the directory in
```

Comment:
`package` is singular not plural. Rename to `packages`? This way `package` and `packages` would be documented separately since they have different criteria.

Reply:
At some point I thought we were avoiding plurals because the english plural rules were confusing to non-english speakers.

I agree it would solve this documentation ambiguity, so can change it.